streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedTasks.java

```diff
@@ -488,20 +500,45 @@ void closeNonAssignedSuspendedTasks(final Map<TaskId, Set<TopicPartition>> newAs
     }
 
     void close(final boolean clean) {
-        close(allTasks(), clean);
-        clear();
-    }
-
-    private void close(final Collection<Task> tasks, final boolean clean) {
-        for (final Task task : tasks) {
+        final AtomicReference<RuntimeException> firstException = new AtomicReference<>(null);
+        for (final Task task : allTasks()) {
             try {
                 task.close(clean, false);
-            } catch (final Throwable t) {
+            } catch (final TaskMigratedException e) {
+                firstException.compareAndSet(null, closeZombieTask(task));
+            } catch (final RuntimeException t) {
+                firstException.compareAndSet(null, t);
                 log.error("Failed while closing {} {} due to the following error:",
                           task.getClass().getSimpleName(),
                           task.id(),
                           t);
+                firstException.compareAndSet(null, closeUncleanIfRequired(task, clean));
```

Comment:
firstException wouldn't be null at this point - see call on line 510
It seems call on line 510 can be removed

Reply:
That is a good catch @tedyu . Thanks for that. cc @mjsax for confirming.

Reply:
I guess, we should keep 510 and change this line to just call `closeUncleanIfRequired` -- would be better to report the first exception back (both get logged but I think the root cause should be thrown and not a follow-up error) Will open follow up PR.

Reply:
See #4046